### ISSUE TITLE
Support for Julia-style `@generated` decorator

### ISSUE THREAD

**Issue by seibert** [numba/numba#1637]
We should integrate this functionality from @jcrist:

https://gist.github.com/jcrist/fddcbb12f3c74748aea5

(similar functionality suggested by @albop and others)

This adds a decorator for creating a CPUDispatcher that, rather than JITing the decorated function when a new type signature is encountered, instead calls the decorated function with the Numba types of the arguments which returns a function to JIT.  Many forms of dynamic code generation require something like this.

**Comment by seibert** [comment]
There is also some discussion as to what the name of the decorator should be:

* `@generated`: Since this feature is basically the same as Julia, we should use their name for it.
* `@generate_jit`: There is some concern that `@generated` looks too much like `@generator` and might be confused with the concept of Python generators.

**Comment by seibert** [comment]
Another thing that crosses my mind: Would we ever want to return function implementations to be compiled with `@vectorize` or `@guvectorize`, rather than `@jit`?  (i.e., would we ever want `@generate_vectorize`?)  This would first require us to unify ufuncs, gufuncs, and JIT functions to all use the same dispatch mechanism.

**Comment by pitrou** [comment]
Well, that would be a possibility.

>  This would first require us to unify ufuncs, gufuncs, and JIT functions to all use the same dispatch mechanism.

That's part of our roadmap :-)

**Comment by pitrou** [comment]
> There is some concern that @generated looks too much like @generator and might be confused with the concept of Python generators

Not only, but it's also a rather vague term. There are tons of things that can be generated :-)

**Comment by pitrou** [comment]
Does someone have a nice use case to test with?

**Comment by seibert** [comment]
This may be broadening the scope a bit, but many of the situations I can think of would also allow the user of `@generated_jit` to specify which arguments should be captured by value when generating the implementation.

For example, implementing `np.sum()` with an axis argument I think requires some code generation that depends on the values.  The interpolator example that prompted the most recent discussion of this has arguments that select particular kinds of interpolation details.

However, I don't think we could get that in this first pass, since it would require our dispatcher to be value-aware, which isn't going to be easy.

**Comment by jcrist** [comment]
Here's actual reason I wrote it:

```python
from generated import generated
import numpy as np
import numba as nb
from dynd import nd

@generated(nopython=True, nogil=True)
def is_missing(x):
    """Returns if the value is missing, per dynd's missing value flag
    semantics"""
    if isinstance(x, nb.types.Array):
        x = x.dtype
    if isinstance(x, nb.types.Integer):
        missing = np.iinfo(x.name).min
        return lambda x: x == missing
    elif isinstance(x, nb.types.Float):
        return lambda x: np.isnan(x)
    else:
        return lambda x: False
    

# Create a dynd array with missing values
x = nd.array([1, None, 2], '3 * ?int32')

# View the internal buffer as a numpy array, which is compatible with numba
buffer = nd.as_numpy(x.view_scalars(x.dtype.value_type))

# Find missing values
out = is_missing(buffer)
# array([False,  True, False], dtype=bool)
```

**Comment by seibert** [comment]
^^^ That is a nice test case.

**Comment by pitrou** [comment]
Oh, yes, thanks.

**Comment by pitrou** [comment]
Your implementation is a bit buggy, though: the last line (`return lambda x: False`) will always return a scalar even if an array is passed.

**Comment by jcrist** [comment]
Oop, you're right. That was a quick last addition for completeness :). Fixed below:

```python
@generated(nopython=True, nogil=True)
def is_missing(x):
    """Returns if the value is missing, per dynd's missing value flag
    semantics"""
    if isinstance(x, nb.types.Array):
        dt = x.dtype
    else:
        dt = x
    if isinstance(dt, nb.types.Integer):
        missing = np.iinfo(dt.name).min
        return lambda x: x == missing
    elif isinstance(dt, nb.types.Float):
        return lambda x: np.isnan(x)
    elif isinstance(x, nb.types.Array):
        return lambda x: np.full_like(x, False)
    else:
        return lambda x: False
```

**Comment by pitrou** [comment]
(note it would also be a nice usecase for a `generated_vectorize` :-))

**Comment by jcrist** [comment]
I wonder if it would be cleaner to make `generated` return an object that the other decorators know about. Then you wouldn't need a `generated_*` for each numba decorator, you'd just stack them:

```
@vectorize(...)
@generated
def foo(x, y):
    ...
```

**Comment by pitrou** [comment]
Implementation wise, it would be more annoying to do so - you have to ensure that the stacking order of decorators is unimportant.

**Comment by pitrou** [comment]
Besides, it's not really a flip you switch (like "nopython"), it really changes how you write your function.

**Comment by albop** [comment]
Since @pitrou was asking for an example, I tried to write a simple one showing how to write a single function for fractional indexing 1d and 2d: https://gist.github.com/albop/d13a58cb2aed98700443. For any array `mat` the goal is to have `findex(mat,i,j)` return `mat[i,j]` when i,j are both integers, and interpolate (linearly) between grid points when they are real numbers.
It doesn't work completely but it highlights a few potential issues:
- in principle, the dispatch mechanism would allow to define `findex(mat,i)` (resp. `findex(mat,i,j)` ) as calls to `findex_1d_(mat,i)` (resp. ` findex_2d(mat,i,j)` ). For some reason, it didn't work, so I ended up dispatching `findex(mat,v)` instead where `v` is a vector. (Lines 140-147 have the original version) 
- When vector v contains integers, one would like to use native indexing. For this I wrote one function of each dimension (`native_index_1d and native_index_2d`). This two routines do the the same thing for two different vector sizes, but the same dimension. This brings the question of whether a particular routine can be optimized based on the length of a vector rather than it's dimension. If I am not mistaken, this would be considered as a value by the current type system. (here it doesn' t matter much because the dispatch is made by the function calling `native_index_1d` but it prevents from writing a unified `native_index` function). This is probably an orthogonal issue.
- This example is potentially another good example for a `guvectorized` generated function, since `guvectorize` works very well with each dimension specific implementation. The main problem, is that the core dimensions are not the same, so that one may want to return them with the generated function (see lines >=152)
